src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -817,6 +877,492 @@ NodeListInsertCommand(List *workerNodeList)
+char *
+DistributedObjectCreateCommand(const ObjectAddress *address,
+							   int32 *distributionArgumentIndex,
+							   int32 *colocationId)
+{
+	StringInfo insertDistributedObjectCommand = makeStringInfo();
+	List *names;
+	List *args;
+	char *objectType = getObjectTypeDescription(address);
+	getObjectIdentityParts(address, &names, &args);
+
+	appendStringInfo(insertDistributedObjectCommand,
+					 "SELECT citus_internal_add_object_metadata(");
+
+	appendStringInfo(insertDistributedObjectCommand,
+					 "%s, ARRAY[",
+					 quote_literal_cstr(objectType));
+
+	char *name;
+	bool firstLoop = true;
+	foreach_ptr(name, names)
+	{
+		if (!firstLoop)
+		{
+			appendStringInfo(insertDistributedObjectCommand, ", ");
+		}
+		firstLoop = false;
+		appendStringInfoString(insertDistributedObjectCommand, quote_literal_cstr(name));
+	}
+
+	appendStringInfo(insertDistributedObjectCommand, "]::text[], ARRAY[");
+
+	char *arg;
+	firstLoop = true;
+	foreach_ptr(arg, args)
+	{
+		if (!firstLoop)
+		{
+			appendStringInfo(insertDistributedObjectCommand, ", ");
+		}
+		firstLoop = false;
+		appendStringInfoString(insertDistributedObjectCommand, quote_literal_cstr(arg));
+	}
+
+	appendStringInfo(insertDistributedObjectCommand, "]::text[],");
+
+	if (distributionArgumentIndex == NULL)
+	{
+		appendStringInfo(insertDistributedObjectCommand, "NULL, ");
+	}
+	else
+	{
+		appendStringInfo(insertDistributedObjectCommand, "%d, ",
+						 *distributionArgumentIndex);
+	}
+	if (colocationId == NULL)
+	{
+		appendStringInfo(insertDistributedObjectCommand, "NULL)");
+	}
+	else
+	{
+		appendStringInfo(insertDistributedObjectCommand, "%d)",
+						 *colocationId);
+	}
+
+	return insertDistributedObjectCommand->data;
+}
+
+
+/*
+ * citus_internal_add_object_metadata is an internal UDF to
+ * add a row to pg_dist_object.
+ */
+Datum
+citus_internal_add_object_metadata(PG_FUNCTION_ARGS)
+{
+	char *textType = TextDatumGetCString(PG_GETARG_DATUM(0));
+	ArrayType *nameArray = PG_GETARG_ARRAYTYPE_P(1);
+	ArrayType *argsArray = PG_GETARG_ARRAYTYPE_P(2);
+	int distributionArgumentIndexValue;
+	int colocationIdValue;
+	int *colocationId = NULL;
+	int *distributionArgumentIndex = NULL;
+
+	if (!PG_ARGISNULL(3))
+	{
+		distributionArgumentIndexValue = PG_GETARG_INT32(3);
+		distributionArgumentIndex = &distributionArgumentIndexValue;
+	}
+	if (!PG_ARGISNULL(4))
+	{
+		colocationIdValue = PG_GETARG_INT32(4);
+		colocationId = &colocationIdValue;
+	}
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+	}
+
+	ObjectAddress objectAddress = GetObjectAddrIfUserOwns(textType, nameArray, argsArray);
+
+	/* Update the metadata for the given object on the node */
+	MarkObjectDistributed(&objectAddress, true);
+	UpdateFunctionDistributionInfo(&objectAddress, distributionArgumentIndex,
+								   colocationId, true);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * Get the object address if user owns it.
```

Comment:
And, I couldn't follow the `user owns it` part. Why do we need that?

(I mean, we need to check if user owns the object, but why in this function? Can we do it seperately?)